lib/ansible/galaxy/dependency_resolution/dataclasses.py

```diff
@@ -39,6 +39,7 @@
 _GALAXY_YAML = b'galaxy.yml'
 _MANIFEST_JSON = b'MANIFEST.json'
 _SOURCE_METADATA_FILE = b'GALAXY.yml'
+SUPPORTED_REQUIREMENTS_KEYS = frozenset({'name', 'version', 'type', 'source', 'signatures'})
```

Comment:
This looks like a good start. I think we really need some validation of the values for these fields as well. Rather than a set of options, this could be written as an argument spec dictionary and passed to `ansible.module_utils.common.arg_spec.ArgumentSpecValidator`. That would give basic type validation, required option validation, and access to normalized parameters for additional validation. These are some of the commonly recurring issues:

* If a requirement is a dictionary, `name` is required. This is the pointer to the collection and accepts the same format as positional collections to install on the CLI. People assume this is just the FQCN/optional.
* `source` is supposed to be a URL to a Galaxy server, not a URL to an artifact. If `name` is not in the format `namespace.collection`, providing `source` should be an error because it's not applicable.
* providing `version` with `type: subdirs` should probably be an error (rather than silently ignored)
* for every other type, except collection in git repos, `version` is required to be a valid semantic version

Reply:
Yea, the validation of these values can be improved on. I'll take a crack at this and see what I can do.